File: jenjin-core/src/main/java/com/binarymonks/jj/physics/PhysicsWorld.java
package com.binarymonks.jj.physics;

import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.physics.box2d.World;
import com.badlogic.gdx.utils.Array;
import com.binarymonks.jj.JJ;
import com.binarymonks.jj.api.Physics;
import com.binarymonks.jj.async.Function;

public class PhysicsWorld implements Physics {

    public World world;
    public int velocityIterations = 10;
    public int positionIterations = 20;
    public CollisionGroups collisionGroups = new CollisionGroupsBasic();
    Array<Function> postPhysicsFunctions = new Array<>();

    boolean updating = false;

    public PhysicsWorld() {
        world = new World(Vector2.Zero, true);
        world.setContactListener(new JJContactListener());
    }

    public void update() {
        updating = true;
        double frameDelta = JJ.time.getDelta();
        if (frameDelta > 0) {
            world.step((float) JJ.time.getDelta(), velocityIterations, positionIterations);
        }
        for (Function function : postPhysicsFunctions) {
            function.call();
        }
        postPhysicsFunctions.clear();
        updating = false;
    }

    public boolean isUpdating() {
        return updating;
    }

    @Override
    public void setCollisionGroups(CollisionGroups collisionGroups) {
        this.collisionGroups = collisionGroups;
    }

    @Override
    public void addPostPhysicsFunction(Function function) {
        postPhysicsFunctions.add(function);
    }
}

File: jenjin-core/src/main/kotlin/com/binarymonks/jj/core/physics/PhysicsWorld.kt
package com.binarymonks.jj.core.physics

import com.badlogic.gdx.physics.box2d.JointDef
import com.badlogic.gdx.physics.box2d.World
import com.binarymonks.jj.core.JJ
import com.binarymonks.jj.core.api.PhysicsAPI
import com.binarymonks.jj.core.async.OneTimeTask
import com.binarymonks.jj.core.pools.Poolable
import com.binarymonks.jj.core.pools.new
import com.binarymonks.jj.core.pools.recycle

open class PhysicsWorld(
        val b2dworld: World = World(JJ.B.config.b2d.gravity, true)
) : PhysicsAPI {


    var velocityIterations = 20
    var positionIterations = 20
    override var collisionGroups = CollisionGroups()
    override var materials = Materials()
    var isUpdating = false
        internal set

    init {
        b2dworld.setContactListener(JJContactListener())
    }

    fun update() {
        isUpdating = true
        val frameDelta = JJ.clock.delta
        if (frameDelta > 0) {
            b2dworld.step(JJ.clock.deltaFloat, velocityIterations, positionIterations)
        }
        isUpdating = false
    }

    /**
     * Lets you create joints during the physics step
     */
    fun createJoint(jointDef: JointDef) {
        if (isUpdating) {
            JJ.tasks.addPostPhysicsTask(new(DelayedCreateJoint::class).set(jointDef))
        } else {
            b2dworld.createJoint(jointDef)
        }
    }

}

internal class DelayedCreateJoint : OneTimeTask(), Poolable {

    var jointDef: JointDef? = null

    fun set(jointDef: JointDef): DelayedCreateJoint {
        this.jointDef = jointDef
        return this
    }

    override fun reset() {
        jointDef = null
    }

    override fun doOnce() {
        JJ.B.physicsWorld.b2dworld.createJoint(jointDef)
        recycle(this)
    }

}
